Chart Builder Revamp updates (#312)
embrace-io/embrace-docs: docs/product/custom-dashboards.md

## Creating a Chart
You can create multiple types of visualizations. A table of which visualizations are supported is shown below:

| Chart Type | Crashes | Network | Logs | Issues | Moments | Sessions | Spans | Historical | Multi-query |
| ---------- | ------- | ------- | ---- | ------ | ------- | -------- | ----- | ---------- | ----------- |
| Line | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |
| Table | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :x:
| Bar | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :x: |
| Pie | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :x: | :x: |

## Spans (Traces)

Create graphs monitoring Spans performance. You can filter and group by name, outcome, duration, and any attributes you have set on the Span.

> - First, select the visualization type you want to see and a source for the first query (more on multi-query combined time series below).

Reviewer: Leave out the reference to the first query if the visualization is open-ended. If we want this particular example to have multiple queries then we can mention the first.

> - First, select the visualization type you want to see and a source for the first query (more on multi-query combined time series below).

Author: ```suggestion
- First, select the visualization type you want to see and a source for the query.
```

- Then, choose the metric you wish to aggregate. For Spans, Embrace supports both counts and sums of duration.
- Finally, add any filters and group-bys. In addition to our core dimensions, for Spans you can use the Span's name, outcome, duration, and any custom Attributes.

<img src={require('@site/static/images/custom_dashboards/Spans_Chart_Builder.png').default} style={{ width: '75%', height: '75%' }} alt="Spans as a Widget option" />

## Combined Time Series (Line Charts)

You can also create Combined graphs in Line vizualizations. With Combined graphs, you can visualize up to 10 time series together and optionally create a custom time series using arithmetic operators.

> **NOTE:** as of 2025-04-29, the chart builder has been improved. Any legacy charts using fills have been moved to multi-query line charts. 

Reviewer: ```suggestion
**NOTE:** as of 2025-05-14, the chart builder has been improved. Any legacy charts using fills have been moved to multi-query line charts. 
```
Hopeful new date 😭

1. Create 2 queries for Network Request, the first filtered for Status Codes in the range 200-299, and the second in the range 400-599.\
    a. You can also add a group by here. For this example, we group by App Version. See more on grouping functionality below. 
    b. Formulas and multiple queries are only supported on line charts currently. 
2. In the Formula box, add the formula `A / (A+B) * 100` to get a successful network request percentage by App Version.\
    a. The formula field accepts the operators: `+, -, *, /`.\
    b. Order of operations follows standard PEMDAS.
3. Hide the queries by clicking on the eye symbols to just show the formula, which is your successful network request percentage.

<img src={require('@site/static/images/custom_dashboards/Formula_Timeseries_Example.png').default} style={{ width: '75%', height: '75%' }} alt="Multiple queries and formula" />

### Multiple Combined Time Series without a Formula

Since the Formula field is optional, you can visualize up to 10 time series on the same chart with the steps above. Use the eye symbols to show or hide each and leave the Formula field blank.

### Grouping with Combined Time Series

> In the broadest sense, the group-by's for each query must have some overlapping set. For example, if you were to set one query to "Crash Count" grouped by "App Version", and the second query to "Session Count", grouped by "Build", then perform some operation on them, this would be an illegal formula.

Author: ```suggestion
In the broadest sense, the group-by's for each query must have some overlapping set. For example, if you were to set one query to "Crash Count" grouped by "App Version", and the second query to "Session Count", grouped by "Build", then perform some operation on them, you will get zero results. An example of this could be:
- `A` grouped by `[app_version]`
- `B` grouped by `[app_version]`
- However `A` has only App Versions 1.0 and 2.0, while `B` has only App Version 3.0. 
```

Additionally, order of operations matters. Consider a situation where you create three queries:
- `A` grouped by `[app_version, os_version]` 
- `B` grouped by `[app_version]` 
- `C` grouped by `[os_version]` 

Then you were to combine them in the formula field:

:white_check_mark: `(A + B) + C`: `A` and `B` share a grouping dimension and can combine, then combine with `C`.\

> :x: `A + (B + C)`: `B` and `C` share now group-by dimension and therefore fail. 

Reviewer: ```suggestion
:x: `A + (B + C)`: `B` and `C` share no group-by dimension and therefore fail. 
```

<img src={require('@site/static/images/custom_dashboards/Grouping_Error.png').default} style={{ width: '75%', height: '75%' }} alt="Example of Grouping Error" />

> 

Reviewer: We should also mention that if we join queries A and B and they have no common labels we will get zero results. So if A has app_version={1.0, 2.0} and app_version={3.0} then they will get zero results.

### Multiple Combined Time Series without a Formula

Since the Formula field is optional, you can visualize up to 10 time series on the same chart with the steps above. Use the eye symbols to show or hide each and leave the Formula field blank.

### Grouping with Combined Time Series

In the broadest sense, the group-by's for each query must have some overlapping set. For example, if you were to set one query to "Crash Count" grouped by "App Version", and the second query to "Session Count", grouped by "Build", then perform some operation on them, you will get zero results. An example of this could be:
- `A` grouped by `[app_version]`
- `B` grouped by `[app_version]`
- However `A` has only App Versions 1.0 and 2.0, while `B` has only App Version 3.0. 

Additionally, order of operations matters. Consider a situation where you create three queries:
- `A` grouped by `[app_version, os_version]` 
- `B` grouped by `[app_version]` 
- `C` grouped by `[os_version]` 

Then you were to combine them in the formula field:

> :white_check_mark: `(A + B) + C`: `A` and `B` share a grouping dimension and can combine, then combine with `C`.\

Author: ```suggestion
- :white_check_mark: `(A + B) + C`: `A` and `B` share a grouping dimension and can combine, then combine with `C`.\
```

> :x: `A + (B + C)`: `B` and `C` share no group-by dimension and therefore fail. 

Author: ```suggestion
- :x: `A + (B + C)`: `B` and `C` share no group-by dimension and therefore fail. 
```
